src/Microsoft.Data/PrimitiveColumnContainer.cs

```diff
@@ -69,91 +84,224 @@ public PrimitiveColumnContainer(long length = 0)
+            }
+            while (count > 0)
+            {
+                if (Buffers.Count == 0)
+                {
+                    Buffers.Add(new DataFrameBuffer<T>());
+                    NullBitMapBuffers.Add(new DataFrameBuffer<byte>());
+                }
+                DataFrameBuffer<T> lastBuffer = Buffers[Buffers.Count - 1];
+                if (lastBuffer.Length == lastBuffer.MaxCapacity)
+                {
+                    lastBuffer = new DataFrameBuffer<T>();
+                    Buffers.Add(lastBuffer);
+                    NullBitMapBuffers.Add(new DataFrameBuffer<byte>());
+                }
+                int allocatable = (int)Math.Min(count, lastBuffer.MaxCapacity);
+                lastBuffer.EnsureCapacity(allocatable);
+                lastBuffer.Span.Slice(lastBuffer.Length, allocatable).Fill(value ?? default);
+                lastBuffer.Length += allocatable;
+                Length += allocatable;
+
+                DataFrameBuffer<byte> lastNullBitMapBuffer = NullBitMapBuffers[NullBitMapBuffers.Count - 1];
+                int nullBitMapAllocatable = (int)Math.Ceiling(allocatable / 8.0);
+                lastNullBitMapBuffer.EnsureCapacity(nullBitMapAllocatable);
+                _modifyNullCountWhileIndexing = false;
+                for (long i = Length - count; i < Length; i++)
+                {
+                    SetValidityBit(i, value.HasValue ? true : false);
+                }
+                _modifyNullCountWhileIndexing = true;
+                lastNullBitMapBuffer.Length += nullBitMapAllocatable;
+                count -= allocatable;
+            }
+        }
+
+        public bool IsValid(long index) => NullCount == 0 || GetValidityBit(index);
+
+        /// <summary>
+        /// A null value has an unset bit
+        /// A NON-null value has a set bit
+        /// </summary>
+        /// <param name="index"></param>
+        /// <param name="value"></param>
+        private void SetValidityBit(long index, bool value)
+        {
+            if ((uint)index > Length)
+            {
+                throw new ArgumentOutOfRangeException(nameof(index));
+            }
+            // First find the right bitMapBuffer
+            int bitMapIndex = (int)(index / Buffers[0].MaxCapacity);
+            Debug.Assert(NullBitMapBuffers.Count > bitMapIndex);
+            DataFrameBuffer<byte> bitMapBuffer = NullBitMapBuffers[bitMapIndex];
+
+            // Set the bit
+            index -= bitMapIndex * Buffers[0].MaxCapacity;
+            int bitMapBufferIndex = (int)((uint)index / 8);
+            Debug.Assert(bitMapBuffer.Length >= bitMapBufferIndex);
+            if (bitMapBuffer.Length == bitMapBufferIndex)
+                bitMapBuffer.Append(0);
+            byte curBitMap = bitMapBuffer[bitMapBufferIndex];
+            byte newBitMap;
+            if (value)
+            {
+                newBitMap = (byte)(curBitMap | (byte)(1 << (int)(index % 8)));
```

Comment:
```suggestion
                newBitMap = (byte)(curBitMap | (byte)(1 << (int)(index & 7)));   // bithack for % 8
```

The same on other places, did not comment them explicitely. Search for `% 8`.